### ISSUE TITLE
ANR seen in com.onesignal.SyncService [com.onesignal.PushRegistratorGPS.isGMSInstalledAndEnabled]

### ISSUE THREAD

**Issue by mitonline** [OneSignal/OneSignal-Android-SDK#267]
Hi,
OneSignal Android SDK version: 3.5.8

We are seeing many ANR's in the console for the service, com.onesignal.SyncService. Please find the stacktrace for all the threads.
```Java
"main" prio=5 tid=1 Native
  | group="main" sCount=1 dsCount=0 obj=0x73ebafa8 self=0x55857c43e0
  | sysTid=985 nice=-4 cgrp=apps sched=0/0 handle=0x7f8e978150
  | state=S schedstat=( 3145216533 1145011635 5219 ) utm=211 stm=103 core=0 HZ=100
  | stack=0x7fce4da000-0x7fce4dc000 stackSize=8MB
  | held mutexes=
  native: pc 000000000005eac4  /system/lib64/libc.so (__ioctl+4)
  native: pc 0000000000068798  /system/lib64/libc.so (ioctl+100)
  native: pc 0000000000027544  /system/lib64/libbinder.so (android::IPCThreadState::talkWithDriver(bool)+164)
  native: pc 0000000000027fa4  /system/lib64/libbinder.so (android::IPCThreadState::waitForResponse(android::Parcel*, int*)+112)
  native: pc 0000000000028218  /system/lib64/libbinder.so (android::IPCThreadState::transact(int, unsigned int, android::Parcel const&, android::Parcel*, unsigned int)+176)
  native: pc 000000000001ff18  /system/lib64/libbinder.so (android::BpBinder::transact(unsigned int, android::Parcel const&, android::Parcel*, unsigned int)+64)
  native: pc 00000000000d9080  /system/lib64/libandroid_runtime.so (???)
  native: pc 0000000000fef95c  /data/dalvik-cache/arm64/system@framework@boot.oat (Java_android_os_BinderProxy_transactNative__ILandroid_os_Parcel_2Landroid_os_Parcel_2I+212)
  at android.os.BinderProxy.transactNative (BinderProxy.java)
  at android.os.BinderProxy.transact (BinderProxy.java:496)
  at android.content.pm.IPackageManager$Stub$Proxy.getPackageInfo (IPackageManager.java:1859)
  at android.app.ApplicationPackageManager.getPackageInfo (ApplicationPackageManager.java:106)
  at com.onesignal.PushRegistratorGPS.isGMSInstalledAndEnabled (PushRegistratorGPS.java:96)
  at com.onesignal.PushRegistratorGPS.internalRegisterForPush (PushRegistratorGPS.java:70)
  at com.onesignal.PushRegistratorGPS.registerForPush (PushRegistratorGPS.java:65)
  at com.onesignal.OneSignal.registerForPushToken (OneSignal.java:458)
  at com.onesignal.OneSignal.makeAndroidParamsRequest (OneSignal.java:487)
  at com.onesignal.OneSignal.startRegistrationOrOnSession (OneSignal.java:435)
  at com.onesignal.OneSignal.onAppFocus (OneSignal.java:712)
  at com.onesignal.ActivityLifecycleHandler.handleFocus (ActivityLifecycleHandler.java:120)
  at com.onesignal.ActivityLifecycleHandler.onActivityResumed (ActivityLifecycleHandler.java:74)
  at com.onesignal.ActivityLifecycleListener.onActivityResumed (ActivityLifecycleListener.java:50)
  at android.app.Application.dispatchActivityResumed (Application.java:208)
  at android.app.Activity.onResume (Activity.java:1235)
  at com.google.android.gms.ads.AdActivity.onResume (AdActivity.java)
  at android.app.Instrumentation.callActivityOnResume (Instrumentation.java:1241)
  at android.app.Activity.performResume (Activity.java:6087)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:2952)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:2994)
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:2373)
  at android.app.ActivityThread.access$800 (ActivityThread.java:144)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1285)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:135)
  at android.app.ActivityThread.main (ActivityThread.java:5233)
  at java.lang.reflect.Method.invoke! (Native method)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:898)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:693)

"Heap thread pool worker thread 0" prio=5 tid=2 Native (still starting up)
  | group="" sCount=1 dsCount=0 obj=0x0 self=0x5585a83500
  | sysTid=989 nice=0 cgrp=apps sched=0/0 handle=0x5585a7d820
  | state=S schedstat=( 2580726 11790159 19 ) utm=0 stm=0 core=0 HZ=100
  | stack=0x7f8aef4000-0x7f8aef6000 stackSize=1020KB
  | held mutexes=
  native: pc 00000000000133cc  /system/lib64/libc.so (syscall+28)
  native: pc 00000000000d0d60  /system/lib64/libart.so (art::ConditionVariable::Wait(art::Thread*)+140)
  native: pc 0000000000318dcc  /system/lib64/libart.so (art::ThreadPoolWorker::Run()+176)
  native: pc 000000000031a2f4  /system/lib64/libart.so (art::ThreadPoolWorker::Callback(void*)+88)
  native: pc 0000000000018a20  /system/lib64/libc.so (__pthread_start(void*)+52)
  native: pc 0000000000014cb0  /system/lib64/libc.so (__start_thread+16)

"Heap thread pool worker thread 2" prio=5 tid=3 Native (still starting up)
  | group="" sCount=1 dsCount=0 obj=0x0 self=0x5585824920
  | sysTid=991 nice=0 cgrp=apps sched=0/0 handle=0x55857d1820
  | state=S schedstat=( 2989584 13216303 20 ) utm=0 stm=0 core=0 HZ=100
  | stack=0x7f8acf2000-0x7f8acf4000 stackSize=1020KB
  | held mutexes=
  native: pc 00000000000133cc  /system/lib64/libc.so (syscall+28)
  native: pc 00000000000d0d60  /system/lib64/libart.so (art::ConditionVariable::Wait(art::Thread*)+140)
  native: pc 0000000000318dcc  /system/lib64/libart.so (art::ThreadPoolWorker::Run()+176)
  native: pc 000000000031a2f4  /system/lib64/libart.so (art::ThreadPoolWorker::Callback(void*)+88)
  native: pc 0000000000018a20  /system/lib64/libc.so (__pthread_start(void*)+52)
  native: pc 0000000000014cb0  /system/lib64/libc.so (__start_thread+16)

"Heap thread pool worker thread 1" prio=5 tid=4 Native (still starting up)
  | group="" sCount=1 dsCount=0 obj=0x0 self=0x55857d1f70
  | sysTid=990 nice=0 cgrp=apps sched=0/0 handle=0x5585a79b50
  | state=S schedstat=( 4024168 8597133 25 ) utm=0 stm=0 core=3 HZ=100
  | stack=0x7f8adf2000-0x7f8adf4000 stackSize=1020KB
  | held mutexes=
  native: pc 00000000000133cc  /system/lib64/libc.so (syscall+28)
  native: pc 00000000000d0d60  /system/lib64/libart.so (art::ConditionVariable::Wait(art::Thread*)+140)
  native: pc 0000000000318dcc  /system/lib64/libart.so (art::ThreadPoolWorker::Run()+176)
  native: pc 000000000031a2f4  /system/lib64/libart.so (art::ThreadPoolWorker::Callback(void*)+88)
  native: pc 0000000000018a20  /system/lib64/libc.so (__pthread_start(void*)+52)
  native: pc 0000000000014cb0  /system/lib64/libc.so (__start_thread+16)

"FinalizerDaemon" daemon prio=5 tid=6 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x12c260e0 self=0x55857cdeb0
  | sysTid=998 nice=0 cgrp=apps sched=0/0 handle=0x55859ccc30
  | state=S schedstat=( 21400729 14881615 54 ) utm=2 stm=0 core=3 HZ=100
  | stack=0x7f867bd000-0x7f867bf000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x285dfc9e> (a java.lang.ref.ReferenceQueue)
  at java.lang.Object.wait (Object.java:422)
  at java.lang.ref.ReferenceQueue.remove (ReferenceQueue.java:101)
- locked <0x285dfc9e> (a java.lang.ref.ReferenceQueue)
  at java.lang.ref.ReferenceQueue.remove (ReferenceQueue.java:72)
  at java.lang.Daemons$FinalizerDaemon.run (Daemons.java:173)
  at java.lang.Thread.run (Thread.java:818)

"HeapTrimmerDaemon" daemon prio=5 tid=7 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x12c261a0 self=0x55859cdb60
  | sysTid=1000 nice=0 cgrp=apps sched=0/0 handle=0x55859ce590
  | state=S schedstat=( 2664844 1961719 16 ) utm=0 stm=0 core=0 HZ=100
  | stack=0x7f7b6f6000-0x7f7b6f8000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x1330e37f> (a java.lang.Daemons$HeapTrimmerDaemon)
  at java.lang.Daemons$HeapTrimmerDaemon.run (Daemons.java:310)
- locked <0x1330e37f> (a java.lang.Daemons$HeapTrimmerDaemon)
  at java.lang.Thread.run (Thread.java:818)

"FinalizerWatchdogDaemon" daemon prio=5 tid=8 Sleeping
  | group="system" sCount=1 dsCount=0 obj=0x12c26140 self=0x55859ccee0
  | sysTid=999 nice=0 cgrp=apps sched=0/0 handle=0x55859cd8b0
  | state=S schedstat=( 1348857 4248021 20 ) utm=0 stm=0 core=0 HZ=100
  | stack=0x7f7b7fa000-0x7f7b7fc000 stackSize=1036KB
  | held mutexes=
  at java.lang.Thread.sleep! (Native method)
- sleeping on <0x045bbf4c> (a java.lang.Object)
  at java.lang.Thread.sleep (Thread.java:1031)
- locked <0x045bbf4c> (a java.lang.Object)
  at java.lang.Thread.sleep (Thread.java:985)
  at java.lang.Daemons$FinalizerWatchdogDaemon.sleepFor (Daemons.java:257)
  at java.lang.Daemons$FinalizerWatchdogDaemon.waitForFinalization (Daemons.java:268)
  at java.lang.Daemons$FinalizerWatchdogDaemon.run (Daemons.java:216)
  at java.lang.Thread.run (Thread.java:818)

"ReferenceQueueDaemon" daemon prio=5 tid=9 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x12c26080 self=0x5585a82d60
  | sysTid=996 nice=0 cgrp=apps sched=0/0 handle=0x55857cdc00
  | state=S schedstat=( 6049015 12316195 41 ) utm=0 stm=0 core=3 HZ=100
  | stack=0x7f8aad2000-0x7f8aad4000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x1021cb95> (a java.lang.Class<java.lang.ref.ReferenceQueue>)
  at java.lang.Daemons$ReferenceQueueDaemon.run (Daemons.java:133)
- locked <0x1021cb95> (a java.lang.Class<java.lang.ref.ReferenceQueue>)
  at java.lang.Thread.run (Thread.java:818)

"GCDaemon" daemon prio=5 tid=10 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x12c26200 self=0x5585a84b80
  | sysTid=1001 nice=0 cgrp=apps sched=0/0 handle=0x5585a85550
  | state=S schedstat=( 508596245 115447770 407 ) utm=49 stm=1 core=3 HZ=100
  | stack=0x7f7b5f2000-0x7f7b5f4000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x34a944aa> (a java.lang.Daemons$GCDaemon)
  at java.lang.Daemons$GCDaemon.run (Daemons.java:341)
- locked <0x34a944aa> (a java.lang.Daemons$GCDaemon)
  at java.lang.Thread.run (Thread.java:818)

"Binder_1" prio=5 tid=11 Native
  | group="main" sCount=1 dsCount=0 obj=0x12c51080 self=0x5585a86bf0
  | sysTid=1002 nice=0 cgrp=apps sched=0/0 handle=0x5585a86780
  | state=S schedstat=( 35648652 71971196 117 ) utm=2 stm=1 core=1 HZ=100
  | stack=0x7f7b3f6000-0x7f7b3f8000 stackSize=1012KB
  | held mutexes=
  native: pc 000000000005eac4  /system/lib64/libc.so (__ioctl+4)
  native: pc 0000000000068798  /system/lib64/libc.so (ioctl+100)
  native: pc 0000000000027544  /system/lib64/libbinder.so (android::IPCThreadState::talkWithDriver(bool)+164)
  native: pc 0000000000027d88  /system/lib64/libbinder.so (android::IPCThreadState::getAndExecuteCommand()+24)
  native: pc 0000000000027e40  /system/lib64/libbinder.so (android::IPCThreadState::joinThreadPool(bool)+76)
  native: pc 000000000002f838  /system/lib64/libbinder.so (???)
  native: pc 0000000000014bb8  /system/lib64/libutils.so (android::Thread::_threadLoop(void*)+208)
  native: pc 000000000008c540  /system/lib64/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+96)
  native: pc 0000000000014404  /system/lib64/libutils.so (???)
  native: pc 0000000000018a20  /system/lib64/libc.so (__pthread_start(void*)+52)
  native: pc 0000000000014cb0  /system/lib64/libc.so (__start_thread+16)

"Binder_2" prio=5 tid=12 Native
  | group="main" sCount=1 dsCount=0 obj=0x12c55080 self=0x5585a88310
  | sysTid=1003 nice=0 cgrp=apps sched=0/0 handle=0x5585a87ce0
  | state=S schedstat=( 30126616 62811614 101 ) utm=0 stm=3 core=1 HZ=100
  | stack=0x7f7b2f8000-0x7f7b2fa000 stackSize=1012KB
  | held mutexes=
  native: pc 000000000005eac4  /system/lib64/libc.so (__ioctl+4)
  native: pc 0000000000068798  /system/lib64/libc.so (ioctl+100)
  native: pc 0000000000027544  /system/lib64/libbinder.so (android::IPCThreadState::talkWithDriver(bool)+164)
  native: pc 0000000000027d88  /system/lib64/libbinder.so (android::IPCThreadState::getAndExecuteCommand()+24)
  native: pc 0000000000027e40  /system/lib64/libbinder.so (android::IPCThreadState::joinThreadPool(bool)+76)
  native: pc 000000000002f838  /system/lib64/libbinder.so (???)
  native: pc 0000000000014bb8  /system/lib64/libutils.so (android::Thread::_threadLoop(void*)+208)
  native: pc 000000000008c540  /system/lib64/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+96)
  native: pc 0000000000014404  /system/lib64/libutils.so (???)
  native: pc 0000000000018a20  /system/lib64/libc.so (__pthread_start(void*)+52)
  native: pc 0000000000014cb0  /system/lib64/libc.so (__start_thread+16)

"Binder_3" prio=5 tid=13 Native
  | group="main" sCount=1 dsCount=0 obj=0x12c5e080 self=0x5585a8b3d0
  | sysTid=1004 nice=0 cgrp=apps sched=0/0 handle=0x5585a89500
  | state=S schedstat=( 19334269 25684689 77 ) utm=1 stm=0 core=1 HZ=100
  | stack=0x7f7b1fa000-0x7f7b1fc000 stackSize=1012KB
  | held mutexes=
  native: pc 000000000005eac4  /system/lib64/libc.so (__ioctl+4)
  native: pc 0000000000068798  /system/lib64/libc.so (ioctl+100)
  native: pc 0000000000027544  /system/lib64/libbinder.so (android::IPCThreadState::talkWithDriver(bool)+164)
  native: pc 0000000000027d88  /system/lib64/libbinder.so (android::IPCThreadState::getAndExecuteCommand()+24)
  native: pc 0000000000027e40  /system/lib64/libbinder.so (android::IPCThreadState::joinThreadPool(bool)+76)
  native: pc 000000000002f838  /system/lib64/libbinder.so (???)
  native: pc 0000000000014bb8  /system/lib64/libutils.so (android::Thread::_threadLoop(void*)+208)
  native: pc 000000000008c540  /system/lib64/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+96)
  native: pc 0000000000014404  /system/lib64/libutils.so (???)
  native: pc 0000000000018a20  /system/lib64/libc.so (__pthread_start(void*)+52)
  native: pc 0000000000014cb0  /system/lib64/libc.so (__start_thread+16)

"FocusHandlerThread" prio=5 tid=14 Native
  | group="main" sCount=1 dsCount=0 obj=0x12c06d60 self=0x5585b98250
  | sysTid=1033 nice=0 cgrp=apps sched=0/0 handle=0x5585b96000
  | state=S schedstat=( 4963175 7982032 26 ) utm=0 stm=0 core=3 HZ=100
  | stack=0x7f79e29000-0x7f79e2b000 stackSize=1036KB
  | held mutexes=
  native: pc 000000000005eb70  /system/lib64/libc.so (__epoll_pwait+8)
  native: pc 0000000000015258  /system/lib64/libc.so (epoll_pwait+32)
  native: pc 0000000000018c48  /system/lib64/libutils.so (android::Looper::pollInner(int)+144)
  native: pc 0000000000019004  /system/lib64/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+76)
  native: pc 00000000000ccee4  /system/lib64/libandroid_runtime.so (android::NativeMessageQueue::pollOnce(_JNIEnv*, int)+48)
  native: pc 000000000012fc30  /data/dalvik-cache/arm64/system@framework@boot.oat (Java_android_os_MessageQueue_nativePollOnce__JI+152)
  at android.os.MessageQueue.nativePollOnce (Native method)
  at android.os.MessageQueue.next (MessageQueue.java:143)
  at android.os.Looper.loop (Looper.java:122)
  at android.os.HandlerThread.run (HandlerThread.java:61)

"OSH_NetworkHandlerThread" prio=5 tid=15 Native
  | group="main" sCount=1 dsCount=0 obj=0x12d90a90 self=0x5585b9a300
  | sysTid=1034 nice=0 cgrp=apps sched=0/0 handle=0x5585b9ad30
  | state=S schedstat=( 58043336 4808281 30 ) utm=0 stm=5 core=2 HZ=100
  | stack=0x7f79d25000-0x7f79d27000 stackSize=1036KB
  | held mutexes=
  native: pc 000000000005eb70  /system/lib64/libc.so (__epoll_pwait+8)
  native: pc 0000000000015258  /system/lib64/libc.so (epoll_pwait+32)
  native: pc 0000000000018c48  /system/lib64/libutils.so (android::Looper::pollInner(int)+144)
  native: pc 0000000000019004  /system/lib64/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+76)
  native: pc 00000000000ccee4  /system/lib64/libandroid_runtime.so (android::NativeMessageQueue::pollOnce(_JNIEnv*, int)+48)
  native: pc 000000000012fc30  /data/dalvik-cache/arm64/system@framework@boot.oat (Java_android_os_MessageQueue_nativePollOnce__JI+152)
  at android.os.MessageQueue.nativePollOnce (Native method)
  at android.os.MessageQueue.next (MessageQueue.java:143)
  at android.os.Looper.loop (Looper.java:122)
  at android.os.HandlerThread.run (HandlerThread.java:61)

"_AsyncTask #1" prio=5 tid=17 Waiting
  | group="main" sCount=1 dsCount=0 obj=0x12c01ca0 self=0x5585c469e0
  | sysTid=1080 nice=10 cgrp=apps/bg_non_interactive sched=0/0 handle=0x5585c5ffa0
  | state=S schedstat=( 19904948 94807342 70 ) utm=1 stm=0 core=0 HZ=100
  | stack=0x7f79ba3000-0x7f79ba5000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x047a389b> (a java.lan

"Signal Catcher" daemon prio=5 tid=5 Runnable
  | group="system" sCount=0 dsCount=0 obj=0x12c00080 self=0x5585a7cd30
  | sysTid=995 nice=0 cgrp=apps sched=0/0 handle=0x55857bce20
  | state=R schedstat=( 187011716 26371305 123 ) utm=5 stm=14 core=0 HZ=100
  | stack=0x7f8abde000-0x7f8abe0000 stackSize=1012KB
  | held mutexes= "thread list lock" "mutator lock"(exclusive held)
  native: pc 0000000000003784  /system/lib64/libbacktrace_libc++.so (Backtrace::Unwind(unsigned long, ucontext*)+28)
  native: pc 0000000000000027  ???
  native: pc 00000000003d1b00  /system/lib64/libart.so (???)
```

**Comment by jkasten2** [comment]
@mitonline I don't see any wait locks on the main thread in this ANR. It's possible something else may have been taken while before it got to this OneSignal stack trace. Do you have any other task traces you can attach here?

Is this ANR from the same app as the #268 ticket you opened?

Lastly I recommend updating to our 3.6.3 SDK as some underlying changes have been made that may have fixed this issue for you.

**Comment by jkasten2** [comment]
We made some large changes to the SyncService which I believe should fix this issue in PR #470. This will released it this week under the version 3.8.2 which you can watch for on our [Releases Page](https://github.com/OneSignal/OneSignal-Android-SDK/releases)

Some ANRs are due to two instances of the main launcher `Activity` being created. Make sure your `Activity` can handle this, or better yet add `android:launchMode="singleTop"` to your `Activity` in your `AndroidManifest.xml` to be safe. See [Google's docs on Actives](https://developer.android.com/guide/topics/manifest/activity-element.html) for more details.
